src/lib/shim/src/lib.rs

```diff
@@ -543,22 +542,19 @@ pub mod export {
 
     /// # Safety
     ///
-    /// Environment variable SHADOW_IPC_BLK must contained a serialized block of
+    /// stdin must contained a serialized block of
     /// type `IPCData`, which outlives the current thread.
     #[no_mangle]
     pub unsafe extern "C" fn _shim_parent_init_ipc() {
-        let envname = CStr::from_bytes_with_nul(b"SHADOW_IPC_BLK\0").unwrap();
-        // SAFETY: Kind of not. We should pass this some other way than an environment
-        // variable. Ok in practice as long as we do our initialization after libc's early
-        // initialization, and nothing else mutates this environment variable.
-        // https://github.com/shadow/shadow/issues/2848
-        let ipc_blk = unsafe { libc::getenv(envname.as_ptr()) };
-        assert!(!ipc_blk.is_null());
-        let ipc_blk = unsafe { CStr::from_ptr(ipc_blk) };
-        let ipc_blk = core::str::from_utf8(ipc_blk.to_bytes()).unwrap();
-
-        use core::str::FromStr;
-        let ipc_blk = ShMemBlockSerialized::from_str(ipc_blk).unwrap();
+        let mut bytes = [0; core::mem::size_of::<ShMemBlockSerialized>()];
+        let bytes_read = rustix::io::read(
+            unsafe { rustix::fd::BorrowedFd::borrow_raw(libc::STDIN_FILENO) },
+            &mut bytes,
+        )
+        .unwrap();
+        // Implement looping? We should get it all in one read, though.
+        assert_eq!(bytes_read, bytes.len());
+        let ipc_blk = unsafe { core::mem::transmute(bytes) };
```

Comment:
I don't think this has the correct alignment. `ShMemBlockSerialized` is also not ffi-safe so may have different layouts in the shim and shadow. Also I don't understand how this works at all since `ShMemBlockSerialized` contains a `PahtBuf` which is dynamically allocated.

Reply:
Oh I think we don't need to worry about alignment. And I see we're using a custom `PahtBuf` which is not dynamically allocated.

I think it would be nicer if the pod crate provided a safe transmute function that we called here instead. Such as `fn transmute(x: impl Pod) -> T`.

Reply:
> I think it would be nicer if the pod crate provided a safe transmute function that we called here instead. Such as `pub fn transmute<T: Pod>(x: impl Pod) -> T`.

Maybe this isn't possible since `Pod` types can have uninitialized memory, so you can't convert a `Pod` to another `Pod`. Though you could probably do `pub fn transmute<const N: usize, T: Pod>(x: [u8; N]) -> T`.

Reply:
Nice; done